### PR TITLE
cloneset partition support intorstr format

### PR DESCRIPTION
**Pull request by hellolijj**
### Ⅰ. Describe what this PR does

cloneset partition support intorstr format

### Ⅱ. Does this pull request fix one issue?

https://github.com/openkruise/kruise/issues/454

### CODE REVIEW

File: apis/apps/v1alpha1/cloneset_types.go [hunk 1]
```diff
@@ -92,7 +92,7 @@ type CloneSetUpdateStrategy struct {
 	// Partition is the desired number of pods in old revisions. It means when partition
 	// is set during pods updating, (replicas - partition) number of pods will be updated.
 	// Default value is 0.
-	Partition *int32 `json:"partition,omitempty"`
+	Partition *intstr.IntOrString `json:"partition,omitempty"`
```
Comment (Fei-Guo) [inline apis/apps/v1alpha1/cloneset_types.go:95]: You need to revise the comment accordingly since it is not a `int` type anymore.

File: pkg/controller/cloneset/scale/cloneset_scale_util_test.go [hunk 2]
```diff
@@ -384,3 +384,7 @@ func TestCalculateDiffs(t *testing.T) {
 		})
 	}
 }
+
+func getIntOrStr(i intstr.IntOrString) *intstr.IntOrString {
```
Comment (Fei-Guo) [inline pkg/controller/cloneset/scale/cloneset_scale_util_test.go:388]: Why do you need this method?

File: pkg/controller/cloneset/cloneset_controller_test.go [hunk 3]
```diff
@@ -396,6 +396,10 @@ func getInt32(i int32) *int32 {
 	return &i
 }
 
+func getIntOrStr(i intstr.IntOrString) *intstr.IntOrString {
```
Comment (Fei-Guo) [inline pkg/controller/cloneset/cloneset_controller_test.go:399]: Why is this needed?

File: pkg/controller/cloneset/update/cloneset_update_test.go [hunk 4]
```diff
@@ -71,6 +71,10 @@ func getInt32Pointer(i int32) *int32 {
 	return &i
 }
 
+func getIntOrStr(i intstrutil.IntOrString) *intstrutil.IntOrString {
```
Comment (Fei-Guo) [inline pkg/controller/cloneset/update/cloneset_update_test.go:74]: same here?

File: apis/apps/v1alpha1/cloneset_types.go [hunk 5]
```diff
@@ -89,10 +89,12 @@ type CloneSetUpdateStrategy struct {
 	// Type indicates the type of the CloneSetUpdateStrategy.
 	// Default is ReCreate.
 	Type CloneSetUpdateStrategyType `json:"type,omitempty"`
-	// Partition is the desired number of pods in old revisions. It means when partition
-	// is set during pods updating, (replicas - partition) number of pods will be updated.
+	// Partition is the desired number of pods in old revisions.
+	// Value can be an absolute number (ex: 5) or a percentage of desired pods (ex: 10%).
+	// Absolute number is calculated from percentage by rounding up by default.
+	//It means when partition is set during pods updating, (replicas - partition value) number of pods will be updated.
```
Comment (Fei-Guo) [inline apis/apps/v1alpha1/cloneset_types.go:95]: Nit:  Add a space between '//' and 'It' in the comment.

**Review by Fei-Guo — COMMENTED** [review]
I have a high level question:

If a user sets 50% as partition and the replica size is 100, this means 50 Pods are in older version. If the user changes the replica size to 200, the partition changes to 100 from controller perspective. I believe in current implementation, controller will not revert 50 Pods to old version. Will this cause confusion? Such confusion does not exist if an absolute value is used in partition.

**Comment by Fei-Guo** [comment]
@hellolijj This is an API change. Can you address my comment above?

**Comment by hellolijj** [comment]
> @hellolijj This is an API change. Can you address my comment above?

Theoretically, this is what we expect, but I need to do some experiments to verify it

**Comment by hellolijj** [comment]
> @hellolijj This is an API change. Can you address my comment above?

here are my [experiments result](https://docs.google.com/document/d/1P4wg2HfSMj8_oFgIxrrRTs0T2MCAjDAMryXL_-0EQew/edit?usp=sharing) about partition change. It's exactly what we want.

**Comment by Fei-Guo** [comment]
@hellolijj Please fix the CI build failure.

**Review by Fei-Guo — APPROVED** [review]
